ledger/ledger.go

```diff
@@ -628,8 +624,6 @@ func (l *Ledger) OnlineCirculation(rnd basics.Round, voteRnd basics.Round) (basi
 
 // CheckDup return whether a transaction is a duplicate one.
 func (l *Ledger) CheckDup(currentProto config.ConsensusParams, current basics.Round, firstValid basics.Round, lastValid basics.Round, txid transactions.Txid, txl ledgercore.Txlease) error {
-	l.trackerMu.RLock()
```

Comment:
I'm a bit worried about removal this lock, all other methods lock. What about a race between close and checkdup?
@cce opinion?

Reply:
Here is my understanding. Please correct me if I am wrong, or tell me if this convinces you. The code should be obvious enough for all to see clearly and understand. Otherwise, we need to revise the design or add more explanation.

After removing `l.trackerMu.RLock()` here, we no longer prevent `CheckDup` and `Ledger.Close` or other trackers `close` from running simultaneously. 

When `l.txTail.checkDup` is called, the `txTail` read lock is grabbed, so any change into the state of `txTail` will be prevented. 
`txTail` does not `close`, it will be garbage collected.

The removal of the lock is a problem if `l.txTail` has references to outside of `txTail` which are modified when `close` is called. 
My answer is that `txTail` is self-contained, and does not reference anything outside its own memory space. So, there should be no problem.

Reply:
sounds right, please add a comment to the inside of `CheckDup` explaining why it is not trackerMu protected as other calls to trackers.

Reply:
I will try to explain in a longer response. In the mean time, if you have any example where the following holds, please tell me, it means I am missing something. 
> we need both tailMu and trackerMu locking

Reply:
> I'm not sure, Ledger.txTail is not a pointer, it is a value, and its initialization is protected by trackerMu ... Ledger.txTail will not be garbage collected, it will be re-initialized e.g. in calls to reloadLedger()

`reloadLedger()` calls `loadFromDisk` on `txTail`, and this call is guarded by `t.tailMu.Lock`. 
So, when `txTail` is re-initialized, no other operation will be taking place on the `txTail` object.

Reply:
> It is very rare for the trackerMu write lock to be acquired, only at close/open time and reloadLedger (catchpoint restore and unit tests) but I thought it was necessary any time you want to access the tracker fields on Ledger, like Ledger.accts, Ledger.acctsOnline, Ledger.catchpoint, Ledger.txTail, etc...

Historically, `txTail` did not have its own mutex. It relied on `trackerMu`. In 320 rounds project #4003 it received its own mutex `tailMu`. However, `tailMu` the coverage of `tailMu` was not complete. `txTail` depended on both `trackerMu` and `tailMu`. This was complicated, unnecessary, and conflicted with the use of txTail header cache in some cases. 

To simplify the logic and unblock me use case, I provided total protection to `txTail` only with `tailMu`, and removed the `trackerMu` dependence.

Regarding other fields of the Ledger you mention here (Ledger.accts, Ledger.acctsOnline, Ledger.catchpoint), they do not share any fields with `txTail`.